lib/controllers/book_controller.dart

```diff
@@ -8,49 +8,123 @@ class LessonState {
+      levels: levels ?? this.levels,
     );
   }
 }
 
 class LessonController extends StateNotifier<LessonState> {
-  LessonController(this._lessonUseCase) : super(const LessonState(loading: false, books: []));
+  LessonController(this._lessonUseCase) : super(const LessonState(loading: false, books: [], subjects: [], levels: []));
   final LessonUseCase _lessonUseCase;
 
   Future<void> fetchInitialBooks() async {
-    print("fetch initial books");
-
     state = state.copyWith(loading: true);
 
     List<BookModel> bookResult = await _lessonUseCase.initBooks();
+
+    bookResult.removeWhere((book) => book.displayTitle == null);
+    List<String> subjects = [];
     for (var book in bookResult) {
-      print(book.url);
+      for (var subject in book.subjects) {
+        if (!subjects.contains(subject)) {
+          subjects.add(subject);
+        }
+      }
     }
-    bookResult.removeWhere((book) => book.displayTitle == null);
+    List<String> levels = [];
+    for (var book in bookResult) {
+      for (var level in book.levels) {
+        if (!levels.contains(level)) {
+          levels.add(level);
+        }
+      }
+    }
+
+    sort(bookResult);
 
     state = state.copyWith(
       loading: false,
       books: bookResult,
+      subjects: subjects,
+      levels: levels,
     );
   }
 
+  void sort(List<BookModel> bookResult) {
+    sortBySubject(bookResult);
+    sortByLevel(bookResult);
+  }
+
+  void sortByLevel(List<BookModel> bookResult) {
+    bookResult.sort((a, b) {
+      String levelA = a.levels.isNotEmpty ? a.levels[0] : '';
+      String levelB = b.levels.isNotEmpty ? b.levels[0] : '';
+      return levelA.compareTo(levelB);
+    });
+  }
+
+  void sortBySubject(List<BookModel> bookResult) {
+    bookResult.sort((a, b) {
+      String subjectA = a.subjects.isNotEmpty ? a.subjects[0] : '';
+      String subjectB = b.subjects.isNotEmpty ? b.subjects[0] : '';
+      return subjectA.compareTo(subjectB);
+    });
+  }
+
   Future<void>? resetAll() {
     state = state.copyWith(
       loading: false,
     );
 
     return null;
   }
+
+  Future<void> fetchBooksByFilter({String? subject, String? level}) async {
+    state = state.copyWith(loading: true);
+
+    List<BookModel> filteredBooks = [];
+    List<BookModel> bookResult = await _lessonUseCase.initBooks();
+    filteredBooks = bookResult;
+    if (subject != null) {
+      filteredBooks = filteredBooks.where((book) => book.subjects.contains(subject)).toList();
+    }
+    if (level != null) {
+      filteredBooks = filteredBooks.where((book) => book.levels.contains(level)).toList();
+    }
+
+    // filteredBooks.sort((a, b) {
+    //   int compare = a.levels[0].compareTo(b.levels[0]);
+    //   if (compare != 0) {
+    //     return compare;
+    //   } else {
+    //     return a.subjects[0].compareTo(b.subjects[0]);
+    //   }
+    // });
+
+    sort(filteredBooks);
+    state = state.copyWith(
+      loading: false,
+      books: filteredBooks,
```

Comment:
It's a bit counter-intuitive that the `books` state depends on values that are themselves not in the state. It makes the state less consistent.